DEVELOPMENT_BEST_PRACTICE.md

```diff
@@ -0,0 +1,145 @@
+
+### Names
+
+##### Packages
+
+Package names help form the first impression of what a package provides and is used to reference the public API of the package. A good package name brings about clarity and purpose for consumers. Package names should be concise and represent the behavior the package provides not the types it contains. For example, `meals` provides `dr
+
+Avoid stutter in struct and interface names by treating the package name as a namespace and considering how the types are consumed. For example, `drink.DrinkDecorator` should be `drink.Dectorator`.
+
+If you find yourself using an appropriate package name that's commonly used as a variable, consider pluralising the package name.
+
+Avoid (like the plague) `types`, `interfaces`, `common`, `util` or `base` packages. They don't represent anything cohesive and tend to collect unrelated types and algorithms that often require teething apart as a project progresses due to cyclic dependencies. If a utility style package is required, make it specific. For example a `/utils/cmdutil` package may provide command utility functions.
+
+##### Function & method names
+
+Functions and methods should adequetly describe their behavior. Generally, they should follow a verb-noun form.
+
+##### Variable names
+
+Variable names should be concise and descriptive. Prefer single word names. The further away from the site of declaration a variable is used, the more descriptive it needs to be. For example, a variable named `n` used 30 lines after declaration makes it unnecessarily difficult to reason what it represents at the site of use.
+
+### Abstractions
+
+- Functions & Methods
+- Constructs
+- Interfaces
+
+### <a nane="dependencies"></a> Depencencies
+
+- Accept interfaces, return structs
+- Declare your dependencies
+
+### Package API
+
+- this might be wrapped up in naming and testing. Do we have more explicit points?
+
+### Testing
+
+- test using the `_test` idiom
+- Use test to help identify if the package is easy to use for consumers
+- isolating concerns like io
+
+### Interfaces
+
+- Define the behavior a _consumer_ expects
+- Premature interfaces
+- Accept interfaces
+- Should be cohesive
+- Should be small
+- Preferably single method
+
+### Errors
+
+- Describe the action that failed
+- Use lower case
+- Don't prepend error
+- Don't overshare information 
+- Add context if you can
+
+### Boolean logic
+
+- Toggling behavior using boolean values is generally bad
+- Create dynamic constructs that can be built without the behavior
+
+### Concurrency
+
+- Prefer synchonous APIs
+- Leave it to the caller
+- Mask from the caller
+
+### Disambiguating context.Context
+
+- Represents the context of execution
+    - Useful for tear down
+- There should be 1 and only 1 context
+- Should be the first parameter on the primary call path
+
+### Loggers
+
+- They're just another dependency, inject them.
```

Comment:
Just an aside I guess, but I see a lot of package-level loggers in k8s code. I sure don't care for that, so I'm glad to see this.

Reply:
Yea I don't personally like the `klog` thing (and some other bits) Kubernetes does. Its the easy way out of managing a significant dependency (the logger). I've worked on projects, and you see it frequently in Go projects, where you let the code grow organically and decide to extract some code to a re-usable repository only to find there are package level dependencies 🤦🏻 .

I'll write as much here as an elaboration to 'why'.